File: src/data/courses/dsa/merge-sort/sorting-comparisons/comparison-vs-non-comparison/code/java/ComparisonVsNonComparison.java
// Comparison-based: Merge Sort - O(n log n)
void mergeSort(int[] arr, int l, int r) {
    if (l < r) {
        int m = l + (r - l) / 2;
        mergeSort(arr, l, m);
        mergeSort(arr, m + 1, r);
        merge(arr, l, m, r);
    }
}

// Non-comparison: Counting Sort - O(n + k)
void countingSort(int[] arr, int k) {
    int[] count = new int[k + 1];
    for (int x : arr) count[x]++;
    int idx = 0;
    for (int i = 0; i <= k; i++) {
        while (count[i]-- > 0) arr[idx++] = i;
    }
}

File: src/data/courses/dsa/merge-sort/sorting-comparisons/comparison-vs-non-comparison/code/python/comparison_vs_non_comparison.py
# Comparison-based: Merge Sort
def merge_sort(arr, l, r):
    if l < r:
        m = l + (r - l) // 2
        merge_sort(arr, l, m)
        merge_sort(arr, m + 1, r)
        merge(arr, l, m, r)

# Non-comparison: Counting Sort
def counting_sort(arr, k):
    count = [0] * (k + 1)
    for x in arr:
        count[x] += 1
    idx = 0
    for i in range(k + 1):
        while count[i] > 0:
            arr[idx] = i
            idx += 1
            count[i] -= 1